[SandboxVec] Add barebones Region class.

slackito opened the pull request:
A region identifies a set of vector instructions generated by vectorization passes. The vectorizer can then run a series of RegionPasses on the region, evaluate the cost, and commit/reject the transforms on a region-by-region basis, instead of an entire basic block.

This is heavily based ov @vporpo's prototype. In particular, the doc comment for the Region class is all his. The rest of this commit is mostly boilerplate around a SetVector: getters, iterators, and some debug helpers.

slackito committed:
[SandboxVec] Add barebones Region class.

A region identifies a set of vector instructions generated by
vectorization passes. The vectorizer can then run a series of
RegionPasses on the region, evaluate the cost, and commit/reject the
transforms on a region-by-region basis, instead of an entire basic
block.

This is heavily based ov @vporpo's prototype. In particular, the doc
comment for the Region class is all his. The rest of this commit is
mostly boilerplate around a SetVector: getters, iterators, and some
debug helpers.

slackito committed:
clang-format

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+/// to check for profitability and decide whether we accept or rollback at a
+/// region granularity, which is much better than doing this at the BB level.
+///
+/// The region keeps track of the costs that correspond to that region. That is
```
I would remove the cost-related comments until we finalize the design.

aeubanks replied:
I'm not a fan of "scalar" vs "vector" since we may be revectorizing existing vector instructions right? It's more like "initial" and "added" cost. But +1 to removing comment for now

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+  /// A unique ID, used for debugging.
+  unsigned RegionID = 0;
+
+  sandboxir::Context &Ctx;
```
It is save to drop the `sandboxir::`.

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+  sandboxir::Context &Ctx;
+
+  /// The basic block containing this region.
+  sandboxir::BasicBlock &SBBB;
```
Rename SBBB to BB

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+  unsigned getID() const { return RegionID; }
+
+  /// Adds SBI to the set.
+  void add(sandboxir::Instruction *SBI);
```
Same here, rename `SBI` to `I`.

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+  Region(sandboxir::Context &Ctx, sandboxir::BasicBlock &SBBB);
+  ~Region();
+
+  sandboxir::BasicBlock *getParent() const { return &SBBB; }
```
Rename `SBBB` to `BB`.

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+
+  /// The cost of the original scalar instructions replaced by vector
+  /// instructions in this region.
+  InstructionCost ScalarCost;
```
Hmm I think we can remove the Cost variables for now. We can add them later if needed.

vporpo reviewed (commented):
Please add a unit test `llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp` to test the API.

Also, regarding the class name, any thoughts on a more suitable name?

tschuett commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,127 @@
+#include "llvm/Support/raw_ostream.h"
+
+namespace llvm {
+namespace sandboxir {
```
You could shorten it to:
```namespace llvm::sandboxir {```

tschuett commented:
The doc is great. Thanks.

slackito committed:
Address review feedback.

- Add tests.
- Remove cost-related comment and unused member variables.
- Drop unneeded "sandboxir::" prefixes.
- Remove "SB" prefix from some variables.
- Simplify nested namespace.
- Added `insts()` method that returns a {begin(), end()} iterator_range.

vporpo commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,107 @@
+class Region {
+  /// All the instructions in the Region. Only new instructions generated during
+  /// vectorization are part of the Region.
+  SetVector<sandboxir::Instruction *> Insts;
```
The `sandboxir::` prefix is not needed because it is the default in the sandboxir namespace. If we need to use an LLVM IR class we would then need to use `llvm::`. So please remove it, it shows up in a couple of places.

slackito replied:
Done. Also moved the code in Region.cpp to the `llvm::sandboxir` namespace.

vporpo commented on llvm/unittests/Transforms/Vectorize/SandboxVectorizer/RegionTest.cpp:
```diff
@@ -0,0 +1,83 @@
+  // that comparison is order-independent.
+  sandboxir::Region Other(Ctx, *BB);
+  Other.add(Ret);
+  Other.add(T1);
```
Could you add a check before you add T1 `EXPECT_NE(Rgn, Other)` ?

slackito replied:
Done. I needed to add an `operator!=` for `EXPECT_NE` (and having only one of the two operators would be confusing for users anyway).

slackito committed:
return of the clang-format

vporpo reviewed (approved):
A couple of small changes, otherwise it looks good to me.

slackito committed:
Another round of review feedback

- Removed the rest of unneeded `sandboxir::` prefixes. Also moved the
  code in Region.cpp to the namespace llvm::sandboxir.

- Added `operator!=` to go with the existing `operator==`, used it in
  the test, as requested.

tmsri commented on llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Region.h:
```diff
@@ -0,0 +1,106 @@
+//  |Rgn2| -> Transform1 ->  ... -> TransformN -> Check Cost
+//  |    |
+//  |Rgn3| -> Transform1 ->  ... -> TransformN -> Check Cost
+//  +----+
```
How do we handle inter-BB transformation?

vporpo replied:
The text is not super accurate, it should be talking about Fn rather than BBs. We will update it.
